<a href="https://colab.research.google.com/github/Shahrukh2016/Logistic_Regression_Implementation/blob/main/LogisticRegression_Gender_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Hi! this is Kaggle dataset obtained from Kaggle.com. I've used this dataset for building the classification Machine Learning Model.**

**It contains 7 features and a label column.**

**longhair** - This column contains 0's and 1's where 1 is "long hair" and 0 is "not long hair".

**foreheadwidthcm** - This column is in CM's. This is the width of the forehead.

**foreheadheightcm** - This is the height of the forehead and it's in Cm's.

**nosewide** - This column contains 0's and 1's where 1 is "wide nose" and 0 is "not wide nose".

**noselong** - This column contains 0's and 1's where 1 is "Long nose" and 0 is "not long nose".

**lipsthin** - This column contains 0's and 1's where 1 represents the "thin lips" while 0 is "Not thin lips".

**distancenosetoliplong** - This column contains 0's and 1's where 1 represents the "long distance between nose and lips" while 0 is "short distance between nose and lips".

**gender** - This is either "Male" or "Female".

"Gender is our target variable"

###**Data Pre-Processing**

In [120]:
# Importing fundamental essentical libraries
import numpy as np
import pandas as pd

In [121]:
# Reading our dataset
df= pd.read_csv("/content/drive/MyDrive/Colab Notebooks/ML Algorithms Shahrukh/Datasets/Gender_Classification.csv")

In [122]:
df.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1    Male  
1          1                          0  Female  
2          1                          1    Male  
3          1                          1    Male  
4          0                          0  Female

First thing first. Lets check our dataset if it has any null, missing, duplicate obsrvations and treat them accordingly.

In [123]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   long_hair                  5001 non-null   int64  
 1   forehead_width_cm          5001 non-null   float64
 2   forehead_height_cm         5001 non-null   float64
 3   nose_wide                  5001 non-null   int64  
 4   nose_long                  5001 non-null   int64  
 5   lips_thin                  5001 non-null   int64  
 6   distance_nose_to_lip_long  5001 non-null   int64  
 7   gender                     5001 non-null   object 
dtypes: float64(2), int64(5), object(1)
memory usage: 312.7+ KB


No null values observed.

In [124]:
# checking duplicated observations
df.duplicated().sum()

1768

In [125]:
# total number of observations
df.shape[0]

5001

Out of 5001 observations we have 1768 duplicated values that makes no sense and will generate unnecessary noice in our predictions. Let's drop them first.


In [126]:
#dropping duplicated observations
df.drop_duplicates(inplace=True)

In [127]:
# rechecking duplicated observations
df.duplicated().sum()

0

Hurray!!! We do not have any null or duplicated value now.

Since this is classification dataset(output variable is discrete/catagorical-Male/Female), to develop and deploy any ML algorithm first we need to format our string variabel to numeric. But before doing that let's just ceate a copy of our dataframe to avoid any blunder in our original dataframe.

In [128]:
#creating copy of our dataframe
df_new= df.copy()

In [129]:
df_new.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1    Male  
1          1                          0  Female  
2          1                          1    Male  
3          1                          1    Male  
4          0                          0  Female

To make our output variable numeric so that it can be trained by any ML Algorithm, let's follow up with the LABEL ENCODING.

###**Label Encoding**

In [130]:
#Formatting Male as 1 and Female as 0
# Importing required library
from sklearn.preprocessing import LabelEncoder

#create instance of label encoder
lab = LabelEncoder()

#perform label encoding on 'team' column
df_new['gender'] = lab.fit_transform(df_new['gender'])

In [131]:
#checking encoded column
df_new.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1       1  
1          1                          0       0  
2          1                          1       1  
3          1                          1       1  
4          0                          0       0

**For the sake of simplicity let's just take our input variable as "long_hair" and output variable as "gender" of course and make the predictions after that we will consider more features and will see if we are gaining more accuracy in our model.** 

###**Train test split(One feature)**

In [132]:
#Assigning observations to our x and y variable
x1= df_new[["long_hair"]]
y1= df_new[["gender"]]

In [133]:
# checking shape of dependent and independent variable
x1.shape, y1.shape

((3233, 1), (3233, 1))

In [134]:
# Importing sklearn library for splitting our data in train and test observations
from sklearn.model_selection import train_test_split

In [135]:
#split operation
x_train, x_test, y_train, y_test = train_test_split(x1, y1, test_size=0.2, random_state=42)

In [136]:
x_train

long_hair
380           1
137           1
2357          1
3215          0
1433          1
...         ...
1282          1
1328          0
1555          1
974           1
4874          0

[2586 rows x 1 columns]

In [137]:
y_train

gender
380        1
137        0
2357       1
3215       1
1433       0
...      ...
1282       1
1328       0
1555       0
974        0
4874       1

[2586 rows x 1 columns]

In [138]:
x_test

long_hair
1187          0
1688          1
1470          1
30            1
141           1
...         ...
765           1
3694          1
900           0
1333          1
1655          1

[647 rows x 1 columns]

In [139]:
y_test

gender
1187       1
1688       0
1470       0
30         1
141        1
...      ...
765        0
3694       0
900        0
1333       1
1655       1

[647 rows x 1 columns]

In [140]:
#checking shape of our splitted dataset
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((2586, 1), (647, 1), (2586, 1), (647, 1))

In [141]:
# checking the distribution ratio of each of catagory in our train dataset 
y_train.value_counts()[1]/y_train.value_counts()[0]

1.2644483362521892

In [142]:
# checking the distribution ratio of each of catagory in our test dataset 
y_test.value_counts()[1]/y_test.value_counts()[0]

1.1006493506493507

Since our catagory distribution ratio is nearly same that means we have balanced dataset in our train and test samples which is good for us. 

###**Model Training(One feature)**

In [143]:
# importing LogisticRegression from sklearn
from sklearn.linear_model import LogisticRegression

In [144]:
# creating instance
classification= LogisticRegression(fit_intercept=True,max_iter=1000000)

In [145]:
#Training the model
classification.fit(x_train,y_train)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(max_iter=1000000)

In [146]:
#checking scoring parameter for our trainins dataset
classification.score(x_train, y_train)

0.5583913379737045

In [147]:
#checing the values of cofficient and intercept
classification.coef_, classification.intercept_

(array([[0.10687628]]), array([0.14710901]))

###**Model Testing(One feature)**

In [148]:
#output train dataset prediction
y_pred_train= classification.predict(x_train)
print(y_pred_train)

[1 1 1 ... 1 1 1]


In [149]:
#output test dataset prediction
y_pred_test= classification.predict(x_test)
print(y_pred_test)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 

In [150]:
# calculating the probabilities of outcume for our training dataset
y_pred_prob_train= classification.predict_proba(x_train)
print(y_pred_prob_train)

[[0.43684283 0.56315717]
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]
 ...
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]
 [0.46328893 0.53671107]]


In [151]:
# calculating the probabilities of outcume for our testing dataset
y_pred_prob_test= classification.predict_proba(x_test)
print(y_pred_prob_test)

[[0.46328893 0.53671107]
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]
 ...
 [0.46328893 0.53671107]
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]]


###**Model Validation(One feature)**

In [152]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import log_loss
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import roc_auc_score

In [153]:
# checking different accuracy metrics for training dataset
print(f"Confusion Matrix is: {confusion_matrix(y_train,y_pred_train)}")
print(f"Accuracy Score is: {accuracy_score(y_train,y_pred_train)*100} %")
print(f"Log loss is: {log_loss(y_train,y_pred_train)}")
print(f"Precision Score is: {precision_score(y_train,y_pred_train)*100} %")
print(f"Recall Score is: {recall_score(y_train,y_pred_train)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_train,y_pred_train)}")

Confusion Matrix is: [[   0 1142]
 [   0 1444]]
Accuracy Score is: 55.83913379737046 %
Log loss is: 15.25297594093321
Precision Score is: 55.83913379737046 %
Recall Score is: 100.0 %
ROC_AUC Score is: 0.5


In [154]:
# checking different accuracy metrics for testing dataset
print(f"Confusion Matrix is: {confusion_matrix(y_test,y_pred_test)}")
print(f"Accuracy Score is: {accuracy_score(y_test,y_pred_test)*100} %")
print(f"Log loss is: {log_loss(y_test,y_pred_test)}")
print(f"Precision Score is: {precision_score(y_test,y_pred_test)*100} %")
print(f"Recall Score is: {recall_score(y_test,y_pred_test)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_test,y_pred_test)}")

Confusion Matrix is: [[  0 308]
 [  0 339]]
Accuracy Score is: 52.39567233384853 %
Log loss is: 16.44233292989334
Precision Score is: 52.39567233384853 %
Recall Score is: 100.0 %
ROC_AUC Score is: 0.5


Since, the accuracy is not much good therefore we have two options with us to improve our accuracy:
1. Try go get more observations(datapoint) which can be use for model training. (Practically not possible in our scenario.)
2. Try to consider more features as input so that model can grasp more information from our training dataset which ultimately results in right prediction and improves accuracy.

Let's follow the second case and explore the output.


##**Its time to feed more input feature and see if our model is improving its accuracy or not.**
I'm very excited to see the upcoming results, and you ??

###**Train test split(Two feature)**

In [155]:
#Assigning observations to our x and y variable
x2= df_new[["long_hair","distance_nose_to_lip_long"]]
y2= df_new[["gender"]]

In [156]:
# checking shape of dependent and independent variable
x2.shape, y2.shape

((3233, 2), (3233, 1))

In [157]:
#split operation
x_train, x_test, y_train, y_test = train_test_split(x2, y2, test_size=0.2, random_state=42)

In [158]:
#checking shape of our splitted dataset
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((2586, 2), (647, 2), (2586, 1), (647, 1))

###**Model Training(Two features)**

In [159]:
# importing LogisticRegression from sklearn
from sklearn.linear_model import LogisticRegression

In [160]:
# creating instance
classification= LogisticRegression(fit_intercept=True,max_iter=1000000)

In [161]:
#Training the model
classification.fit(x_train,y_train)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(max_iter=1000000)

In [162]:
#checking scoring parameter for our trainins dataset
classification.score(x_train, y_train)

0.8194122196442382

In [163]:
#checing the values of cofficient and intercept
classification.coef_, classification.intercept_

(array([[0.22909563, 2.98328425]]), array([-1.50762386]))

###**Model Testing(Two feature)**

In [164]:
#output train dataset prediction
y_pred_train= classification.predict(x_train)
print(y_pred_train)

[0 0 1 ... 1 0 1]


In [165]:
#output test dataset prediction
y_pred_test= classification.predict(x_test)
print(y_pred_test)

[0 0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 0 1 0 1 1 1 1 0 1 0 1 0 0 0 1 0 1 1 1 0 1
 0 1 0 1 1 1 0 0 0 0 1 1 1 1 1 0 0 1 0 1 0 0 1 0 1 1 0 0 1 0 0 0 0 1 1 0 0
 1 0 0 1 0 0 0 1 0 1 0 0 1 0 0 1 0 1 0 0 0 1 1 0 1 0 1 1 0 1 1 0 1 1 0 1 1
 1 1 1 1 0 0 0 1 1 1 1 1 0 0 1 1 1 0 0 1 0 0 0 0 1 1 0 0 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 0 1 0 0 0 1 1 0 0 0 0 1 0 0 0 1 0 0 0 1 0 0 1 0 1 1 1 1 0 0 0 1
 1 1 0 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 0 0 1 0 1 0 1 1 1 0 0 1 1 0 0 1 1 1 1
 1 0 0 0 0 1 0 1 1 1 1 0 1 1 0 1 1 1 0 1 1 1 1 0 1 0 1 1 0 1 0 1 0 0 0 1 1
 0 1 1 1 0 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 0 1 1 0 1 1 0 1 0 1 0 1 1 1 1 1
 1 1 1 1 0 0 0 1 0 0 1 1 0 0 0 0 0 0 1 0 0 0 1 1 0 0 1 1 0 1 0 1 1 1 1 0 0
 0 1 1 1 1 0 1 1 0 0 1 0 0 1 1 0 1 0 0 0 0 1 1 1 0 1 0 0 1 0 0 0 1 0 0 0 0
 1 0 0 1 0 1 0 1 1 1 1 0 0 1 0 1 0 0 1 0 1 0 1 1 0 0 1 0 1 0 1 0 0 1 1 0 1
 0 0 0 0 1 0 0 1 0 0 1 0 0 1 1 1 1 0 1 0 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 1 0
 0 1 1 0 0 1 1 1 1 1 1 1 1 0 0 0 1 0 0 1 0 0 1 1 0 1 1 1 0 1 1 0 1 0 1 1 1
 0 0 0 1 1 1 0 1 0 0 0 1 

In [166]:
# calculating the probabilities of outcume for our training dataset
y_pred_prob_train= classification.predict_proba(x_train)
print(y_pred_prob_train)

[[0.78219915 0.21780085]
 [0.78219915 0.21780085]
 [0.15384512 0.84615488]
 ...
 [0.15384512 0.84615488]
 [0.78219915 0.21780085]
 [0.18608379 0.81391621]]


In [167]:
# calculating the probabilities of outcume for our testing dataset
y_pred_prob_test= classification.predict_proba(x_test)
print(y_pred_prob_test)

[[0.8187088  0.1812912 ]
 [0.78219915 0.21780085]
 [0.78219915 0.21780085]
 ...
 [0.8187088  0.1812912 ]
 [0.78219915 0.21780085]
 [0.15384512 0.84615488]]


###**Model Validation(Two feature)**

In [168]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import log_loss
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import roc_auc_score

In [169]:
# checking different accuracy metrics for training dataset
print(f"Confusion Matrix is: {confusion_matrix(y_train,y_pred_train)}")
print(f"Accuracy Score is: {accuracy_score(y_train,y_pred_train)*100} %")
print(f"Log loss is: {log_loss(y_train,y_pred_train)}")
print(f"Precision Score is: {precision_score(y_train,y_pred_train)*100} %")
print(f"Recall Score is: {recall_score(y_train,y_pred_train)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_train,y_pred_train)}")

Confusion Matrix is: [[ 917  225]
 [ 242 1202]]
Accuracy Score is: 81.94122196442382 %
Log loss is: 6.237350535902971
Precision Score is: 84.23265592151367 %
Recall Score is: 83.2409972299169 %
ROC_AUC Score is: 0.8176936026119314


In [170]:
# checking different accuracy metrics for testing dataset
print(f"Confusion Matrix is: {confusion_matrix(y_test,y_pred_test)}")
print(f"Accuracy Score is: {accuracy_score(y_test,y_pred_test)*100} %")
print(f"Log loss is: {log_loss(y_test,y_pred_test)}")
print(f"Precision Score is: {precision_score(y_test,y_pred_test)*100} %")
print(f"Recall Score is: {recall_score(y_test,y_pred_test)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_test,y_pred_test)}")

Confusion Matrix is: [[260  48]
 [ 51 288]]
Accuracy Score is: 84.69860896445131 %
Log loss is: 5.284972556063071
Precision Score is: 85.71428571428571 %
Recall Score is: 84.95575221238938 %
ROC_AUC Score is: 0.8468566831398688


Hurray!!!

We got some interesting results: As our **input features increases**, our **accuracy is also increases**.


##**Let's consider all the features given in our dataset and try to maximize the accuracy**

###**Train test split(All feature)**

In [171]:
#Assigning observations to our x and y variable
xf= df_new[["long_hair","forehead_width_cm","forehead_height_cm","nose_wide",	"nose_long","lips_thin","distance_nose_to_lip_long"]]
yf= df_new[["gender"]]

In [172]:
# checking shape of dependent and independent variable
xf.shape, yf.shape

((3233, 7), (3233, 1))

In [173]:
#split operation
x_train, x_test, y_train, y_test = train_test_split(xf, yf, test_size=0.2, random_state=42)

In [174]:
#checking shape of our splitted dataset
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((2586, 7), (647, 7), (2586, 1), (647, 1))

###**Model Training(All features)**

In [175]:
# creating instance
classification= LogisticRegression(fit_intercept=True,max_iter=1000000)

In [176]:
#Training the model
classification.fit(x_train,y_train)
     

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(max_iter=1000000)

In [177]:
#checking scoring parameter for our trainins dataset
classification.score(x_train, y_train)

0.9501160092807425

In [178]:
#checing the values of cofficient and intercept
classification.coef_, classification.intercept_

(array([[-0.05919102,  0.61587526,  1.00504779,  3.63070658,  3.29433343,
          3.37061782,  3.30704316]]), array([-20.73540636]))

###**Model Testing(All feature)**

In [179]:
#output train dataset prediction
y_pred_train= classification.predict(x_train)
print(y_pred_train)

[1 0 1 ... 0 0 1]


In [180]:
#output test dataset prediction
y_pred_test= classification.predict(x_test)
print(y_pred_test)

[0 0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 0 1 1 1 1 1 1 0 1 0 1 0 0 0 1 0 1 1 1 0 1
 0 1 0 1 1 1 0 0 0 0 1 1 1 1 1 0 0 1 0 1 1 0 1 0 1 1 0 0 1 1 0 0 0 1 1 0 1
 0 1 0 1 0 0 0 1 0 1 1 0 1 0 0 1 0 1 0 1 0 1 1 0 1 0 1 1 0 1 0 0 1 1 0 0 1
 1 1 0 1 0 0 0 1 1 1 1 0 0 0 1 1 1 1 0 1 1 0 0 1 1 1 1 0 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 0 1 0 0 1 1 1 0 1 0 1 0 0 0 0 1 0 0 0 1 0 0 1 0 1 1 1 1 0 0 0 0
 1 1 0 0 0 1 0 0 0 0 1 0 1 1 1 0 0 1 0 1 1 0 1 0 1 0 1 0 0 1 1 0 0 1 1 0 0
 1 0 0 0 0 1 1 0 1 1 1 0 1 1 0 0 1 1 0 1 1 1 1 0 0 0 1 0 1 0 0 1 0 0 0 1 1
 0 1 1 1 0 0 1 1 1 1 1 1 0 0 1 1 0 1 1 0 0 0 1 1 0 0 1 1 1 0 1 0 1 0 1 1 1
 1 1 1 0 0 0 1 1 0 1 1 0 1 0 0 0 0 0 1 0 1 0 1 1 0 0 1 0 0 1 1 1 1 1 0 1 0
 0 0 1 1 1 0 1 1 0 1 1 1 0 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0
 0 0 0 1 0 1 0 1 1 1 1 0 0 1 0 0 0 0 1 0 0 0 1 1 0 0 1 0 1 0 1 0 0 1 0 0 1
 0 0 0 0 1 0 0 1 0 0 1 1 1 1 1 1 1 1 0 0 1 1 1 0 1 1 1 0 1 1 1 0 1 1 0 0 1
 0 1 1 0 0 0 1 1 1 0 1 1 1 1 0 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 0 1 0 1 0 0
 0 1 0 1 0 1 0 1 0 0 1 1 

In [181]:
# calculating the probabilities of outcume for our training dataset
y_pred_prob_train= classification.predict_proba(x_train)
print(y_pred_prob_train)

[[0.07720706 0.92279294]
 [0.5359952  0.4640048 ]
 [0.08369892 0.91630108]
 ...
 [0.97298267 0.02701733]
 [0.99860459 0.00139541]
 [0.13481505 0.86518495]]


In [182]:
# calculating the probabilities of outcume for our testing dataset
y_pred_prob_test= classification.predict_proba(x_test)
print(y_pred_prob_test)

[[5.49819337e-01 4.50180663e-01]
 [5.95818507e-01 4.04181493e-01]
 [9.64619230e-01 3.53807697e-02]
 ...
 [9.99143537e-01 8.56462547e-04]
 [4.53667547e-01 5.46332453e-01]
 [2.96887674e-04 9.99703112e-01]]


###**Model Validation(All feature)**

In [186]:
#importing essential libraries to check the accuracy
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import log_loss
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import roc_auc_score

In [184]:
# checking different accuracy metrics for training dataset
print(f"Confusion Matrix is: {confusion_matrix(y_train,y_pred_train)}")
print(f"Accuracy Score is: {accuracy_score(y_train,y_pred_train)*100} %")
print(f"Log loss is: {log_loss(y_train,y_pred_train)}")
print(f"Precision Score is: {precision_score(y_train,y_pred_train)*100} %")
print(f"Recall Score is: {recall_score(y_train,y_pred_train)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_train,y_pred_train)}")

Confusion Matrix is: [[1076   66]
 [  63 1381]]
Accuracy Score is: 95.01160092807424 %
Log loss is: 1.7229524084972438
Precision Score is: 95.4388389771942 %
Recall Score is: 95.6371191135734 %
ROC_AUC Score is: 0.9492889230634889


In [185]:
# checking different accuracy metrics for testing dataset
print(f"Confusion Matrix is: {confusion_matrix(y_test,y_pred_test)}")
print(f"Accuracy Score is: {accuracy_score(y_test,y_pred_test)*100} %")
print(f"Log loss is: {log_loss(y_test,y_pred_test)}")
print(f"Precision Score is: {precision_score(y_test,y_pred_test)*100} %")
print(f"Recall Score is: {recall_score(y_test,y_pred_test)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_test,y_pred_test)}")

Confusion Matrix is: [[296  12]
 [ 14 325]]
Accuracy Score is: 95.98145285935085 %
Log loss is: 1.387971841478888
Precision Score is: 96.43916913946587 %
Recall Score is: 95.87020648967551 %
ROC_AUC Score is: 0.9598705129678581


Superb, we have achieved **95% accuracy score** by including all our features as an input. Although this model can be improved further by performing some **feature engineering** techniques and other methods too. but in order to keep things **simple and precise** we can say that this model is good enough and can be proceed further for **deployment**.
